### ISSUE TITLE
Feature: merging keras and tf.keras frameworks

### ISSUE THREAD

**Issue by qubvel** [qubvel/efficientnet#39]
Hi @Callidior
I am going to adapt my `segmentation-models` lib for tf.keras, so I also need efficientnet adapted for tf.keras.
I saw your implementation of Efficientnet for keras-applications. Very nice work, but it is too long to wait until keras will merge it and make a new release.
It would be nice if you make a PR with same code to this repo, or allow to copy it.

**Comment by Callidior** [comment]
Hi @qubvel 
I am glad that you find my contribution to `keras-applications` useful and I would be happy to help you. I am, however, not exactly sure how I can help you, since you already seem to have quite a nice implementation of EfficientNet. Could you provide me with more details?

**Comment by qubvel** [comment]
@Callidior My model is not suited both for keras and tf.keras, I have two different branches for that. But your model written in keras applications style to work with both. If you want you can just paste your model code to https://github.com/qubvel/efficientnet/blob/feature/merging-keras-tf.keras/efficientnet/model.py (`feature/merging-keras-tf.keras` branch) and make a PR. I will do the rest. Or I can copy it by myself, if you are not against it, and mention you as author.

**Comment by Callidior** [comment]
Okay, I created PR #42 with an exact copy of my code. Hope it helps!

**Comment by qubvel** [comment]
Everything works ok for keras, but tf.keras throw an arrow in Dropout layer https://travis-ci.com/qubvel/efficientnet/jobs/218365138
@Callidior do you have any thoughts?

**Comment by Callidior** [comment]
In my opinion, this is a bug in `tf.keras`. I compared the implementation of the Dropout layer with that of `keras` and found that `tf.keras` does not replace `None`-valued dimensions with the actual batch size. `keras` does this in [`keras.layers.Dropout._get_noise_shape`](https://github.com/keras-team/keras/blob/master/keras/layers/core.py#L110):

```python
def _get_noise_shape(self, inputs):
    if self.noise_shape is None:
        return self.noise_shape

    symbolic_shape = K.shape(inputs)
    noise_shape = [symbolic_shape[axis] if shape is None else shape
                   for axis, shape in enumerate(self.noise_shape)]
    return tuple(noise_shape)
```

`tf.keras` just leaves the dimension of the batch axis as `None`, which leads to the error.

A clean fix would be to derive a custom dropout layer compatible with both frameworks, which implements the correct behaviour:
```python
class Dropout(layers.Dropout):
    
    def _get_noise_shape(self, inputs):
        
        if self.noise_shape is None:
            return self.noise_shape

        symbolic_shape = backend.shape(inputs)
        noise_shape = [symbolic_shape[axis] if shape is None else shape
                       for axis, shape in enumerate(self.noise_shape)]
        return tuple(noise_shape)
```

But then, you need to include this layer in the `custom_objects` dictionary for de-serializing saved models.

A more hacky fix, which I didn't completely test, would be to replace `noise_shape=(None, 1, 1, 1)` in the model construction code with `noise_shape=(backend.shape(x)[0], 1, 1, 1)`, but I don't like this solution and suspect that it might lead to different errors at some point.

**Comment by qubvel** [comment]
Thanks for clarification, I dont like that I have to shadow new Dropout class definition, because at first time `layers` module is None and I can not use inheritance from `layers.Layer`. But may be its a price for compatibility)
Are you going to make a bug report to `tensorflow`?

I also have included `swish` to global custom objects to make model deserializable. May be you also need to do it in your keras-applications PR?

**Comment by Callidior** [comment]
I don't like any of these workarounds either and have file a bug report at https://github.com/tensorflow/tensorflow/issues/30946, but who knows how long it will take them to fix it.

Maybe the second "hack" I suggested might be a quick solution for you in the meantime.

**Comment by qubvel** [comment]
Hack with `noise_shape=(backend.shape(x)[0], 1, 1, 1)` does not work, because object is not serializable, while defining a custom Dropout layer works fine. Thanks a lot :+1:

**Comment by Callidior** [comment]
Hm, I somewhat expected that. Thanks for checking. :+1: 
But besides the custom Dropout layer, I don't see another way at the moment, unfortunately.

**Comment by Callidior** [comment]
It was noted in the tensorflow issue that this bug has recently been fixed in https://github.com/tensorflow/tensorflow/commit/ab2644e7178312fc9c89f90d7c8e61f73f7c1f64. But that commit does not seem to have made it into a release yet.

**Comment by qubvel** [comment]
Thanks, will wait for release to update)
For now the work is done, thanks again for your help!
